File: legacy/packages/runtime/src/mock-llm.ts
import type { JsonSchema } from "@moon/schema-compiler";
import type { LlmClient, LlmRequest } from "./llm-types";
import { LlmValidationError } from "./llm-types";
import type { MetricsCollector } from "./metrics";
import type { PricingTable } from "./pricing";
import { countTokens } from "./tokenizer";

export { LlmValidationError, type LlmClient, type LlmRequest } from "./llm-types";

export function fixtureFromSchema(schema: JsonSchema, path = "root"): unknown {
  if (schema.enum && schema.enum.length > 0) {
    return schema.enum[0];
  }

  switch (schema.type) {
    case "string":
      return `mock-${path}`;
    case "integer":
      return 1;
    case "number":
      return schema.minimum ?? 0.5;
    case "boolean":
      return true;
    case "array":
      return schema.items ? [fixtureFromSchema(schema.items, `${path}[0]`)] : [];
    case "object": {
      const obj: Record<string, unknown> = {};
      if (schema.properties) {
        for (const [key, prop] of Object.entries(schema.properties)) {
          obj[key] = fixtureFromSchema(prop, `${path}.${key}`);
        }
      }
      return obj;
    }
    default:
      return { _mock: path };
  }
}

export function createMockLlm(overrides?: {
  responses?: Record<string, unknown>;
  invalidAgents?: Set<string>;
  maxRetries?: number;
  metrics?: MetricsCollector;
  pricing?: PricingTable;
}): LlmClient & { callCount: number; retries: number } {
  let callCount = 0;
  let retries = 0;
  const maxRetries = overrides?.maxRetries ?? 1;

  return {
    callCount: 0,
    retries: 0,
    async complete(req: LlmRequest): Promise<unknown> {
      callCount++;
      const key = `${req.agent}:${JSON.stringify(req.input)}`;
      if (overrides?.responses?.[key]) {
        return overrides.responses[key];
      }
      if (overrides?.responses?.[req.agent]) {
        return overrides.responses[req.agent]!;
      }

      if (overrides?.invalidAgents?.has(req.agent) && retries < maxRetries) {
        retries++;
        throw new LlmValidationError("Mock invalid response");
      }

      const output = fixtureFromSchema(req.schema);
      if (overrides?.metrics && overrides?.pricing) {
        const promptText = JSON.stringify({
          system: req.systemPrompt,
          input: req.input,
          config: req.config,
        });
        const outputText = JSON.stringify(output);
        overrides.metrics.recordLlmUsage(req.model, {
          prompt: countTokens(promptText, req.model, overrides.pricing),
          completion: countTokens(outputText, req.model, overrides.pricing),
          cacheHit: 0,
          cacheMiss: countTokens(promptText, req.model, overrides.pricing),
        }, 1);
      }
      return output;
    },
  };
}

File: Sources/MoonRuntime/MockLLM.swift
import Foundation

public func fixtureFromSchema(_ schema: JsonSchema, path: String = "root") -> RuntimeValue {
    switch schema {
    case .string(let enumValues):
        if let first = enumValues?.first { return .string(first) }
        return .string("mock-\(path)")
    case .integer:
        return .int(1)
    case .number(let minimum, _):
        return .double(minimum ?? 0.5)
    case .boolean:
        return .bool(true)
    case .array(let items):
        if let items {
            return .array([fixtureFromSchema(items, path: "\(path)[0]")])
        }
        return .array([])
    case .object(let properties, _):
        var fields: [String: RuntimeValue] = [:]
        if let properties {
            for (key, prop) in properties {
                fields[key] = fixtureFromSchema(prop, path: "\(path).\(key)")
            }
        }
        return .record(typeName: nil, fields: fields)
    }
}

public final class MockLlmClient: LlmClient, @unchecked Sendable {
    public private(set) var callCount: Int = 0

    public init() {}

    public func complete(_ request: LlmRequest) async throws -> RuntimeValue {
        callCount += 1
        return fixtureFromSchema(request.schema)
    }
}